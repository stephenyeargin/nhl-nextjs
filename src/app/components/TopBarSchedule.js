'use client';

import React, { useState, useEffect } from 'react'
import Image from 'next/image';
import Link from 'next/link';
import dayjs from 'dayjs';
import timezone from 'dayjs/plugin/timezone';
import utc from 'dayjs/plugin/utc';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBan, faWarning } from '@fortawesome/free-solid-svg-icons';
import { PERIOD_DESCRIPTORS } from '../utils/constants';
import { formatGameTime } from '../utils/formatters';

dayjs.extend(timezone);
dayjs.extend(utc);

const sortGamesByState = (games) => {
  const sortedGames = [...games].sort((a, b) => {
    if (a.gameState === 'LIVE' || a.gameState === 'CRIT') return -1;
    if (b.gameState === 'LIVE' || b.gameState === 'CRIT') return 1;
    return 0;
  });
  return sortedGames;
}

const TopBarSchedule = ({ gameDate }) => {
  const [games, setGames] = useState([]);
  const [focusedDate, setFocusedDate] = useState(gameDate || null);
  const [dates, setDates] = useState([]);

  // Fetch the game data on initial load and when focusedDate changes
  useEffect(() => {
    const fetchGames = async () => {
      try {
        const response = await fetch('/api/nhl/scoreboard/now', { cache: 'reload' });
        const data = await response.json();

        // If focusedDate is not set initially, set it from the API response
        if (!focusedDate) {
          setFocusedDate(gameDate || data.focusedDate || data.gamesByDate[0]?.date);
        }

        // Set the available dates from the API response
        setDates(data.gamesByDate.map((d) => d.date));

        // Find the games for the focused date
        const gamesForFocusedDate = data.gamesByDate.find((g) => focusedDate === g.date)?.games || [];
        setGames(sortGamesByState(gamesForFocusedDate));
      } catch (error) {
        console.error('Error fetching games:', error);
      }
    };

    fetchGames();
  }, [focusedDate, gameDate]); // Re-run when focusedDate or gameDate changes

  // Set up the interval to fetch games every 30 seconds
  useEffect(() => {
    const interval = setInterval(() => {
      const fetchGames = async () => {
        try {
          const response = await fetch('/api/nhl/scoreboard/now', { cache: 'reload' });
          const data = await response.json();
          const gamesForFocusedDate = data.gamesByDate.find((g) => focusedDate === g.date)?.games || [];
          setGames(sortGamesByState(gamesForFocusedDate));
        } catch (error) {
          console.error('Error fetching games on interval:', error);
        }
      };

      fetchGames();
    }, 30000); // Refresh every 30 seconds

    return () => clearInterval(interval); // Cleanup interval on component unmount
  }, [focusedDate]); // Only re-run interval if focusedDate changes

  // Handle date click to update focusedDate
  const handleDateClick = (date) => {
    setFocusedDate(date);
  };

  return (
    <div>
      {games?.length > 0 ? (
        <>
          <div className="flex text-xs py-4">
            {dates.map((date) => {
              let dateClass = 'mx-1 border rounded-xl';
              if (dayjs(date).format('YYYY-MM-DD') === dayjs().format('YYYY-MM-DD')) {
                dateClass = 'mx-1 border border-blue-400 rounded-xl'
              }
              if (date === focusedDate) {
                dateClass = 'active rounded-xl bg-slate-500 text-white mx-1'
              }
              return (
                <Link key={date} href={`?date=${date}`} className={dateClass} onClick={() => handleDateClick(date)}>
                  <div className="px-4 text-center">{dayjs(date).utc().format('MMM D')}</div>
                </Link>
              )
              })}
          </div>
          <div className="overflow-x-auto scrollbar-hidden">
            <div className="flex flex-nowrap gap-4 p-4">
              {games.map((game) => {

                game.awayTeam.defeated = game.awayTeam.score < game.homeTeam.score && ['FINAL', 'OFF'].includes(game.gameState);
                game.homeTeam.defeated = game.homeTeam.score < game.awayTeam.score && ['FINAL', 'OFF'].includes(game.gameState);

                let itemClass = 'border rounded-lg shadow-sm p-4 hover:shadow-md transition-shadow';
                if (game.gameState === 'CRIT') {
                  itemClass = 'border border-red-500 rounded-lg shadow-sm p-4 hover:shadow-md transition-shadow';
                }

                return (
                  <Link key={game.id} href={`/game/${game.id}`} className={itemClass} style={{minWidth: '380px'}}>
                    <div className="space-y-2">
                      {/* Away Team */}
                      <div className={`flex items-center justify-between ${game.awayTeam.defeated ? 'opacity-50' : ''}`}>
                        <div className="flex items-center space-x-3">
                          <Image 
                            src={game.awayTeam.logo} 
                            alt={`${game.awayTeam.name?.default} logo`}
                            className="w-8 h-8"
                            width={32}
                            height={32}
                          />
                          <span className="font-bold">
                            {game.awayTeam.name?.default}
                            {game.situation?.awayTeam.situationDescriptions?.map((code) => (
                              <span key={code} className="text-sm font-medium px-2 py-1 bg-red-900 text-white rounded ml-2">{code}</span>
                            ))}
                          </span>
                        </div>
                        <span className="text-lg font-semibold">
                          {game.gameState !== 'FUT' ? game.awayTeam.score : ''}
                        </span>
                      </div>

                      {/* Home Team */}
                      <div className={`flex items-center justify-between ${game.homeTeam.defeated ? 'opacity-50' : ''}`}>
                        <div className="flex items-center space-x-3">
                          <Image 
                            src={game.homeTeam.logo} 
                            alt={`${game.homeTeam.name?.default} logo`}
                            className="w-8 h-8"
                            width={32}
                            height={32}
                          />
                          <span className="font-bold">
                            {game.homeTeam.name?.default}
                            {game.situation?.homeTeam.situationDescriptions?.map((code) => (
                              <span key={code} className="text-sm font-medium px-2 py-1 bg-red-900 text-white rounded ml-2">{code}</span>
                            ))}
                          </span>
                        </div>
                        <span className="text-lg font-semibold">
                          {game.gameState !== 'FUT' ? game.homeTeam.score : ''}
                        </span>
                      </div>
                    </div>
                    <div className="mt-2 pt-3 border-t">
                      <div className="flex justify-between items-center">
                        <span className="text-sm text-slate-600">{game.venue.default}</span>
                        {game.gameScheduleState === 'CNCL' && (
                          <span className="text-xs font-medium px-2 py-1 bg-slate-900 text-white rounded mr-1 uppercase"><FontAwesomeIcon icon={faBan} fixedWidth /> Cancelled</span>
                        )}
                        {game.gameScheduleState === 'PPD' && (
                          <span className="text-xs font-medium px-2 py-1 bg-yellow-500 text-black rounded mr-1 uppercase"><FontAwesomeIcon icon={faWarning} fixedWidth /> Postponed</span>
                        )}
                        {(game.gameState === 'FINAL' || game.gameState === 'OFF') ? (
                          <span className="text-xs font-medium px-2 py-1 bg-slate-100 dark:text-black rounded">
                            FINAL{game.periodDescriptor?.periodType !== 'REG' ? `/${game.periodDescriptor?.periodType}` : ''}
                          </span>
                        ) : (
                          <>
                          {(game.gameState === 'LIVE' || game.gameState === 'CRIT') ? (
                            <span className="text-sm">
                              <span className="text-xs font-medium px-2 py-1 bg-red-900 text-white rounded mr-2 uppercase">
                                {PERIOD_DESCRIPTORS[game.periodDescriptor.number]}
                                {game.clock.inIntermission ? ' INT' : ''}
                              </span>
                              <span className="font-bold">{game.clock?.timeRemaining}</span>
                            </span>
                          ) : (
                            <>
                            <span className="text-sm">
                              {formatGameTime(game.startTimeUTC)}
                            </span>
                            </>
                          )}
                          </>
                        )}
                      </div>
                    </div>
                  </Link>
                );
              })}
            </div>
          </div>
        </>
      ) : (
        <div className="overflow-x-auto scrollbar-hidden">
          <div className="flex text-sm py-4">
            {[0,1,2,3,4,5,6].map((i) => (
              <div key={i} className={i == 3 ? 'active rounded-xl bg-slate-500 mx-3' : 'active rounded-xl bg-slate-100 mx-3'}>
                <div className="px-4">&nbsp;</div>
              </div>
            ))}
          </div>
          <div className="flex flex-nowrap gap-4 p-4">
            {[0,1,3,4,5,6].map((placeholder) => (
              <div key={placeholder} className="border rounded-lg shadow-sm p-4 hover:shadow-md transition-shadow" style={{minWidth: '380px'}}>
                <div className="space-y-2">
                  <div className="flex items-center justify-between">
                    <div className="mx-3 text-sm bg-slate-700 w-10">&nbsp;</div>
                    <div className="mx-3 text-sm bg-slate-700 w-80">&nbsp;</div>
                  </div>
                  <div className="flex items-center justify-between">
                    <div className="mx-3 text-sm bg-slate-700 w-10">&nbsp;</div>
                    <div className="mx-3 text-sm bg-slate-700 w-80">&nbsp;</div>
                  </div>
                </div>
                <div className="mt-2 pt-3">
                  <div className="flex justify-between items-center">
                    <div className="mx-3 text-sm bg-slate-700 w-80">&nbsp;</div>
                  </div>
                </div>
              </div>
            ))}
          </div>
        </div>
      )}
    </div>
  );
}

export default TopBarSchedule;
