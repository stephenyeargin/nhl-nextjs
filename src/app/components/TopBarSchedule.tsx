'use client';

import React, { useState, useEffect } from 'react';
import dayjs from 'dayjs';
import timezone from 'dayjs/plugin/timezone';
import utc from 'dayjs/plugin/utc';
import GameTile from './GameTile';
import TopBarScheduleSkeleton from './TopBarScheduleSkeleton';
import { formatLocalizedDate } from '../utils/formatters';

dayjs.extend(timezone);
dayjs.extend(utc);

interface ScheduleGameTeam {
  abbrev: string;
  score: number;
  record?: string;
  logo?: string;
  placeName?: { default: string };
  commonName?: { default: string };
  id?: string | number;
}
interface ScheduleGame {
  id: string | number;
  gameState: string;
  gameScheduleState?: string;
  gameType: number;
  awayTeam: ScheduleGameTeam;
  homeTeam: ScheduleGameTeam;
  startTimeUTC: string;
  seriesStatus?: any;
  gameOutcome?: { lastPeriodType?: string };
  periodDescriptor?: any;
  clock?: { timeRemaining?: string; inIntermission?: boolean };
  ifNecessary?: boolean;
  venue?: { default: string };
}

const sortGamesByState = (games: ScheduleGame[]): ScheduleGame[] => {
  const sortedGames = [...games].sort((a, b) => {
    if (a.gameState === 'LIVE' || a.gameState === 'CRIT') {
      return -1;
    }
    if (b.gameState === 'LIVE' || b.gameState === 'CRIT') {
      return 1;
    }

    return 0;
  });

  return sortedGames;
};

interface TopBarScheduleProps {
  gameDate?: string;
}
const TopBarSchedule: React.FC<TopBarScheduleProps> = ({ gameDate }) => {
  const [games, setGames] = useState<ScheduleGame[] | null>(null);
  const [focusedDate, setFocusedDate] = useState<string | null>(gameDate || null);
  const [dates, setDates] = useState<string[]>([]);

  // Fetch the game data on initial load and when focusedDate changes
  useEffect(() => {
    const fetchGames = async () => {
      try {
        const response = await fetch('/api/nhl/scoreboard/now', { cache: 'reload' });
        const data = await response.json();
        // Establish initial target date preference order
        let targetDate =
          focusedDate || gameDate || data.focusedDate || data.gamesByDate[0]?.date || null;

        // Set list of all dates provided by API (even those without games)
        setDates(data.gamesByDate.map((d: any) => d.date));

        const findGames = (date: string | null) =>
          date ? data.gamesByDate.find((g: any) => date === g.date)?.games || [] : [];

        let gamesForTarget = findGames(targetDate);

        // If no games for the chosen date, find the closest date with games.
        if (targetDate && gamesForTarget.length === 0) {
          const candidates = data.gamesByDate.filter(
            (d: any) => Array.isArray(d.games) && d.games.length > 0
          );

          // Helper to compute closest date (prefers later date when equidistant)
          if (candidates.length > 0) {
            const pivot = dayjs(targetDate);
            let best = candidates[0];
            let bestDiff = Math.abs(dayjs(best.date).diff(pivot, 'day'));
            for (const c of candidates) {
              const diff = Math.abs(dayjs(c.date).diff(pivot, 'day'));
              if (diff < bestDiff) {
                best = c;
                bestDiff = diff;
              } else if (diff === bestDiff && dayjs(c.date).isAfter(dayjs(best.date))) {
                // Tie -> choose the later calendar date
                best = c;
              }
            }
            if (best.date !== targetDate) {
              targetDate = best.date;
              setFocusedDate(best.date); // trigger highlight update
              gamesForTarget = best.games; // use these games immediately (avoid flash)
            }
          }
        } else if (!focusedDate && targetDate) {
          // If we just derived an initial date (with games) set it.
          setFocusedDate(targetDate);
        }

        setGames(sortGamesByState(gamesForTarget));
      } catch (error) {
        console.error('Error fetching games:', error);
      }
    };

    fetchGames();
  }, [focusedDate, gameDate]); // Re-run when focusedDate or gameDate changes

  // Set up the interval to fetch games every 30 seconds
  useEffect(() => {
    const interval = setInterval(() => {
      const fetchGames = async () => {
        try {
          const response = await fetch('/api/nhl/scoreboard/now', { cache: 'reload' });
          const data = await response.json();
          const gamesForFocusedDate =
            data.gamesByDate.find((g: any) => focusedDate === g.date)?.games || [];
          if (gamesForFocusedDate.length > 0) {
            setGames(sortGamesByState(gamesForFocusedDate));

            return;
          }
          // Fallback again if the currently focused date now has zero games (e.g., data update)
          const candidates = data.gamesByDate.filter(
            (d: any) => Array.isArray(d.games) && d.games.length > 0
          );
          if (candidates.length > 0 && focusedDate) {
            const pivot = dayjs(focusedDate);
            let best = candidates[0];
            let bestDiff = Math.abs(dayjs(best.date).diff(pivot, 'day'));
            for (const c of candidates) {
              const diff = Math.abs(dayjs(c.date).diff(pivot, 'day'));
              if (diff < bestDiff) {
                best = c;
                bestDiff = diff;
              } else if (diff === bestDiff && dayjs(c.date).isAfter(dayjs(best.date))) {
                best = c;
              }
            }
            if (best.date !== focusedDate) {
              setFocusedDate(best.date);
              setGames(sortGamesByState(best.games));

              return;
            }
            setGames(sortGamesByState(best.games));
          } else {
            setGames([]);
          }
        } catch (error) {
          console.error('Error fetching games on interval:', error);
        }
      };

      fetchGames();
    }, 30000); // Refresh every 30 seconds

    return () => clearInterval(interval); // Cleanup interval on component unmount
  }, [focusedDate]); // Only re-run interval if focusedDate changes

  // Handle date click to update focusedDate
  const handleDateClick = (date: string) => {
    setFocusedDate(date);
  };

  if (!games) {
    return <TopBarScheduleSkeleton />;
  }

  return (
    <div className="px-2 my-3">
      <div className="flex text-xs gap-2 overflow-x-auto scrollbar-hidden">
        {dates.map((date) => {
          let dateClass = 'border rounded-xl';
          if (dayjs(date).format('YYYY-MM-DD') === dayjs().format('YYYY-MM-DD')) {
            dateClass = 'border border-blue-400 rounded-xl';
          }
          if (date === focusedDate) {
            dateClass = 'active rounded-xl bg-slate-500 text-white';
          }

          return (
            <button
              key={date}
              className={dateClass}
              onClick={() => handleDateClick(date)}
              title={formatLocalizedDate(date, 'dddd, MMMM D, YYYY')}
            >
              <div className="px-4 text-center">{dayjs(date).utc().format('MMM D')}</div>
            </button>
          );
        })}
      </div>
      <div className="overflow-x-auto scrollbar-hidden my-3">
        <div className="flex flex-nowrap gap-4">
          {games.length === 0 && (
            <div
              className="flex items-center border rounded"
              style={{ minHeight: '9.25rem', minWidth: '360px' }}
            >
              <div className="p-4 text-gray-500">No games scheduled for today.</div>
            </div>
          )}
          {games.map((game) => (
            <GameTile
              key={game.id}
              game={game as any}
              hideDate={true}
              style={{ minWidth: '360px' }}
            />
          ))}
        </div>
      </div>
    </div>
  );
};

export default TopBarSchedule;
