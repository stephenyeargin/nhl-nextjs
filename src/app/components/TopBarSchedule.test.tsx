import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import TopBarSchedule from './TopBarSchedule';

// Mock GameTile to lightweight placeholder
jest.mock('./GameTile', () => {
  // Destructure the actual prop name `hideDate` and rename to `_hideDate` to omit it from DOM spread
  const GT = ({ hideDate: _hideDate, ...props }: any) => (
    <div data-testid="game-tile" {...props}>
      Game {props.game?.id}
    </div>
  );
  (GT as any).displayName = 'GameTileMock';

  return GT;
});

const makeApiPayload = () => ({
  focusedDate: '2024-10-01',
  gamesByDate: [
    {
      date: '2024-10-01',
      games: [
        {
          id: 1,
          gameState: 'FUT',
          awayTeam: { abbrev: 'A', score: 0 },
          homeTeam: { abbrev: 'H', score: 0 },
          startTimeUTC: '2024-10-01T23:00:00Z',
          gameType: 2,
        },
      ],
    },
    {
      date: '2024-10-02',
      games: [
        {
          id: 2,
          gameState: 'FUT',
          awayTeam: { abbrev: 'B', score: 0 },
          homeTeam: { abbrev: 'H', score: 0 },
          startTimeUTC: '2024-10-02T23:00:00Z',
          gameType: 2,
        },
      ],
    },
  ],
});

describe('TopBarSchedule (smoke)', () => {
  beforeEach(() => {
    global.fetch = jest.fn().mockResolvedValue({ json: async () => makeApiPayload() });
  });
  afterEach(() => {
    (global.fetch as jest.Mock).mockReset();
  });

  it('renders dates and switches focused date', async () => {
    const { asFragment } = render(<TopBarSchedule gameDate="2024-10-01" />);
    await waitFor(() => expect(screen.getByText(/Oct 1/i)).toBeInTheDocument());
    expect(screen.getByTestId('game-tile')).toHaveTextContent('Game 1');
    expect(asFragment()).toMatchSnapshot();
    fireEvent.click(screen.getByTitle(/October 2/i));
    await waitFor(() => expect(screen.getByTestId('game-tile')).toHaveTextContent('Game 2'));
    expect(asFragment()).toMatchSnapshot();
  });
});
