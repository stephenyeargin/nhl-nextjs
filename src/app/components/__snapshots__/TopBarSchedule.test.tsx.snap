// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBarSchedule (smoke) renders dates and switches focused date 1`] = `
<DocumentFragment>
  <div
    class="px-2 my-3"
  >
    <div
      class="flex text-xs gap-2 overflow-x-auto scrollbar-hidden"
    >
      <button
        class="active rounded-xl bg-slate-500 text-white"
        title="Tuesday, October 1, 2024"
      >
        <div
          class="px-4 text-center"
        >
          Oct 1
        </div>
      </button>
      <button
        class="border rounded-xl"
        title="Wednesday, October 2, 2024"
      >
        <div
          class="px-4 text-center"
        >
          Oct 2
        </div>
      </button>
    </div>
    <div
      class="overflow-x-auto scrollbar-hidden my-3"
    >
      <div
        class="flex flex-nowrap gap-4"
      >
        <div
          data-testid="game-tile"
          game="[object Object]"
          style="min-width: 360px;"
        >
          Game 1
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TopBarSchedule (smoke) renders dates and switches focused date 2`] = `
<DocumentFragment>
  <div
    class="px-2 my-3"
  >
    <div
      class="flex text-xs gap-2 overflow-x-auto scrollbar-hidden"
    >
      <button
        class="border rounded-xl"
        title="Tuesday, October 1, 2024"
      >
        <div
          class="px-4 text-center"
        >
          Oct 1
        </div>
      </button>
      <button
        class="active rounded-xl bg-slate-500 text-white"
        title="Wednesday, October 2, 2024"
      >
        <div
          class="px-4 text-center"
        >
          Oct 2
        </div>
      </button>
    </div>
    <div
      class="overflow-x-auto scrollbar-hidden my-3"
    >
      <div
        class="flex flex-nowrap gap-4"
      >
        <div
          data-testid="game-tile"
          game="[object Object]"
          style="min-width: 360px;"
        >
          Game 2
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
